### ISSUE TITLE
Rule proposal: no-redundant-array-map

### ISSUE THREAD

**Issue by Jessidhia** [eslint/eslint#12168]
**Please describe what the rule should do:**
The `Array.from` method, introduced in ES2015, has a second argument that is used for mapping the input values to the produced array. This means that, in the majority of cases (where the 3rd argument to the `Array#map` callback is not used), using `Array.from(...).map(...)` is a redundancy that creates a useless intermediate array.

The rule should warn about and autofix `Array.from(Expression).map((arg1, arg2, arg3) => Body)` (where arg3 is either not present or is unused) to `Array.from(Expression, (arg1, arg2) => Body)`.

**What category of rule is this? (place an "X" next to just one item)**

[ ] Warns about a potential error (problem)
[x] Suggests an alternate way of doing something (suggestion)
[ ] Enforces code style (layout)
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

```js
// Warnings
Array.from(foo).map(a => a.bar)
Array.from(bar).map((_, i) => i)
Array.from(baz).map((_, i, arr) => i)

// OK
Array.from(foo)
Array.from(foo, a => a.bar)
Array.from(bar, (_, i) => i)
Array.from({ length: 5 }, (_, i) => i + 1)
Array.from(baz).map((_, i, arr) => i > 0 ? arr[i - 1] + 1 : 0)
```

Example on real code, like this snippet from https://developers.google.com/web/updates/2017/09/sticky-headers
```diff
-return Array.from(container.querySelectorAll('.sticky')).map(el => {
+return Array.from(container.querySelectorAll('.sticky'), el => {
   const sentinel = document.createElement('div');
   sentinel.classList.add('sticky_sentinel', className);
   return el.parentElement.appendChild(sentinel);
 })
```

**Why should this rule be included in ESLint (instead of a plugin)?**
It's about a core API in ES2015+ javascript, rather than anything environment-specific. It warns about a misuse of the API that's not necessarily dangerous, but is redundant and unlikely to ever happen other than by mistake.

**Are you willing to submit a pull request to implement this rule?**
Yes

**Comment by platinumazure** [comment]
Hi @Jessidhia, thanks for the suggestion.

I think we would want the rule name to very clearly indicate the `Array#map` function rather than the `Map` constructor, so maybe `no-redundant-array-map`. But others might have even better name suggestions. :smile:

I'm more or less okay with seeing this as a core rule, but we have to maintain a high bar for new rules because we already have way too many. I'm not 100% convinced this use case is a critical issue worth the maintenance cost of adding a new rule. But other team members may disagree.

**Comment by ilyavolodin** [comment]
I think this is borderline case. It's generic enough to be in the core. It also suggests best practice. But it feels like an edge case that will not be triggered too often. Hmm... not sure which way I'm leaning. Would want to see what others think.

**Comment by kaicataldo** [comment]
Though I see value in this rule, it feels a little too specific to be in core. I would advocate for this being in a plugin!
